nixos/modules/services/web-apps/peertube.nix

```diff
@@ -0,0 +1,355 @@
+        default = true;
+        description = "Configure local Postfix SMTP server for PeerTube";
+      };
+    };
+
+    package = lib.mkOption {
+      type = lib.types.package;
+      default = pkgs.peertube;
+      description = "Peertube package to use";
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.tmpfiles.rules = [
+      "d '/var/lib/peertube/config' 0700 ${cfg.user} ${cfg.group} - -"
+      "z '/var/lib/peertube/config' 0700 ${cfg.user} ${cfg.group} - -"
+      "d '/var/lib/peertube/storage' 0750 ${cfg.user} ${cfg.group} - -"
+      "z '/var/lib/peertube/storage' 0750 ${cfg.user} ${cfg.group} - -"
+    ];
+
+    systemd.services.peertube-init-db = lib.mkIf databaseActuallyCreateLocally {
+      description = "Initialization database for PeerTube daemon";
+      after = [ "network.target" "postgresql.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      script = ''
+        ${config.services.postgresql.package}/bin/psql -c "CREATE EXTENSION IF NOT EXISTS pg_trgm;" ${cfg.database.name}
+        ${config.services.postgresql.package}/bin/psql -c "CREATE EXTENSION IF NOT EXISTS unaccent;" ${cfg.database.name}
+      '';
+
+      serviceConfig = {
+        Type = "oneshot";
+        WorkingDirectory = cfg.package;
+        # User and group
+        User = "postgres";
+        Group = "postgres";
+        # Sandboxing
+        RestrictAddressFamilies = [ "~AF_INET" "~AF_INET6" "~AF_NETLINK" "~AF_PACKET" ];
+        MemoryDenyWriteExecute = true;
+        # System Call Filtering
+        SystemCallFilter = "~@clock @cpu-emulation @debug @keyring @memlock @module @mount @obsolete @privileged @raw-io @reboot @resources @setuid @swap";
+      } // cfgService;
+    };
+
+    systemd.services.peertube = {
+      description = "PeerTube daemon";
+      after = [ "network.target" ]
+        ++ lib.optionals redisActuallyCreateLocally [ "redis.service" ]
+        ++ lib.optionals databaseActuallyCreateLocally [ "postgresql.service" "peertube-init-db.service" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment.NODE_CONFIG_DIR = "/var/lib/peertube/config";
+      environment.NODE_ENV = "production";
+      environment.NODE_EXTRA_CA_CERTS = "/etc/ssl/certs/ca-certificates.crt";
+      environment.HOME = cfg.package;
+
+      path = with pkgs; [ bashInteractive ffmpeg nodejs openssl yarn youtube-dl ];
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStartPre = let preStartScript = pkgs.writeScript "peertube-pre-start.sh" ''
+          #!/bin/sh
+          umask 077
+          cat > /var/lib/peertube/config/local.yaml <<EOF
+          ${lib.optionalString ((!databaseActuallyCreateLocally) && (cfg.database.passwordFile != null)) ''
+          database:
+            password: '$(cat ${cfg.database.passwordFile})'
+          ''}
+          ${lib.optionalString ((redisActuallyCreateLocally) && (cfg.redis.passwordFile == null)) ''
+          redis:
+            hostname:
```

Comment:
what about setting the values in redis.host, redis.port, redis.socket conditionally on createLocally? Then this could be simplified to just check for passwordFile (and maybe password and auth could be set unconditionally but I don't know if that works)?

Reply:
The default value `hostname: 'localhost'` has a higher priority than `socket: '/run/redis/redis.sock'`. The value `hostname:` need to clear.